waku/v2/protocol/waku_store/rpc_codec.nim

```diff
@@ -33,180 +33,222 @@ proc encode*(index: PagingIndexRPC): ProtoBuffer =
+  else:
+    let pagingInfo = ?PagingInfoRPC.decode(pagingInfoBuffer)
+    rpc.pagingInfo = some(pagingInfo)
 
   var startTime: zint64
-  discard ?pb.getField(5, startTime)
-  msg.startTime = Timestamp(startTime)
+  if not ?pb.getField(5, startTime):
+    rpc.startTime = none(int64)
+  else:
+    rpc.startTime = some(int64(startTime))
 
   var endTime: zint64
-  discard ?pb.getField(6, endTime)
-  msg.endTime = Timestamp(endTime)
+  if not ?pb.getField(6, endTime):
+    rpc.endTime = none(int64)
+  else:
+    rpc.endTime = some(int64(endTime))
 
-  ok(msg)
+  ok(rpc)
 
 
 proc encode*(response: HistoryResponseRPC): ProtoBuffer =
   var pb = initProtoBuffer()
 
-  for msg in response.messages:
-    pb.write3(2, msg.encode())
+  for rpc in response.messages:
+    pb.write3(2, rpc.encode())
 
-  pb.write3(3, response.pagingInfo.encode())
+  pb.write3(3, response.pagingInfo.map(encode))
   pb.write3(4, uint32(ord(response.error)))
   pb.finish3()
 
   pb
 
 proc decode*(T: type HistoryResponseRPC, buffer: seq[byte]): ProtoResult[T] =
-  var msg = HistoryResponseRPC()
+  var rpc = HistoryResponseRPC()
   let pb = initProtoBuffer(buffer)
 
   var messages: seq[seq[byte]]
-  discard ?pb.getRepeatedField(2, messages)
-
-  for pb in messages:
-    let message = ?WakuMessage.decode(pb)
-    msg.messages.add(message)
+  if ?pb.getRepeatedField(2, messages):
+    for pb in messages:
+      let message = ?WakuMessage.decode(pb)
+      rpc.messages.add(message)
+  else:
+    rpc.messages = @[]
 
   var pagingInfoBuffer: seq[byte]
-  discard ?pb.getField(3, pagingInfoBuffer)
-  msg.pagingInfo = ?PagingInfoRPC.decode(pagingInfoBuffer)
+  if ?pb.getField(3, pagingInfoBuffer):
+    let pagingInfo = ?PagingInfoRPC.decode(pagingInfoBuffer)
+    rpc.pagingInfo = some(pagingInfo)
+  else:
+    rpc.pagingInfo = none(PagingInfoRPC)
 
   var error: uint32
-  discard ?pb.getField(4, error)
-  msg.error = HistoryResponseErrorRPC.parse(error)
+  if not ?pb.getField(4, error):
+    return err(ProtoError.RequiredFieldMissing)
+  else:
+    rpc.error = HistoryResponseErrorRPC.parse(error)
 
-  ok(msg)
+  ok(rpc)
 
 
 proc encode*(rpc: HistoryRPC): ProtoBuffer =
   var pb = initProtoBuffer()
 
   pb.write3(1, rpc.requestId)
-  pb.write3(2, rpc.query.encode())
-  pb.write3(3, rpc.response.encode())
+  pb.write3(2, rpc.query.map(encode))
+  pb.write3(3, rpc.response.map(encode))
   pb.finish3()
 
   pb
 
 proc decode*(T: type HistoryRPC, buffer: seq[byte]): ProtoResult[T] =
   var rpc = HistoryRPC()
   let pb = initProtoBuffer(buffer)
-  discard ?pb.getField(1, rpc.requestId)
+
+  if not ?pb.getField(1, rpc.requestId):
+    return err(ProtoError.RequiredFieldMissing)
 
   var queryBuffer: seq[byte]
-  discard ?pb.getField(2, queryBuffer)
-  rpc.query = ?HistoryQueryRPC.decode(queryBuffer)
+  if not ?pb.getField(2, queryBuffer):
+    rpc.query = none(HistoryQueryRPC)
+  else:
+    let query = ?HistoryQueryRPC.decode(queryBuffer)
+    rpc.query = some(query)
```

Comment:
I see this structure repeated a little bit in the decoding procs. I know we aren't supposed to use templates extensively, but this looks like a good place to use them. Doesn't have to be a part of this PR though :)

Reply:
First, establish the pattern, then extract and apply the template. This is how I usually create "extensible code" 😉

>  but this looks like a good place to use them.

Here we are using `libp2p`'s protobuf library. I don't know yet if we'll stick with it or if we'll move to `nim-protobuf-serialization` (see #1361).

Reply:
Sounds good, thanks!